src/Facebook/FileUpload/FacebookResumableUploader.php

```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * @var string
+     */
+    protected $accessToken;
+
+    /**
+     * @var FacebookClient The Facebook client service.
+     */
+    protected $client;
+
+    /**
+     * @var string Graph version to use for this request.
+     */
+    protected $graphVersion;
+
+    /**
+     * @param FacebookApp $app
+     * @param FacebookClient $client
+     * @param string $accessToken
+     * @param string $graphVersion
+     */
+    public function __construct(FacebookApp $app, FacebookClient $client, $accessToken, $graphVersion)
+    {
+        $this->app = $app;
+        $this->client = $client;
+        $this->accessToken = $accessToken;
+        $this->graphVersion = $graphVersion;
+    }
+
+    /**
+     * Upload by chunks - start phase
+     *
+     * @param string $endpoint
+     * @param FacebookFile $file
+     *
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookSDKException
+     */
+    public function start($endpoint, FacebookFile $file)
+    {
+        $params = [
+            'upload_phase' => 'start',
+            'file_size' => $file->getSize(),
+        ];
+        $response = $this->sendUploadRequest($endpoint, $params);
+
+        return new FacebookTransferChunk($file, $response['upload_session_id'], $response['video_id'], $response['start_offset'], $response['end_offset']);
+    }
+
+    /**
+     * Upload by chunks - transfer phase
+     *
+     * @param string $endpoint
+     * @param FacebookTransferChunk $chunk
+     * @param boolean $allowToThrow
+     *
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookResponseException
+     */
+    public function transfer($endpoint, FacebookTransferChunk $chunk, $allowToThrow = false)
+    {
+        $params = [
+            'upload_phase' => 'transfer',
+            'upload_session_id' => $chunk->getUploadSessionId(),
+            'start_offset' => $chunk->getStartOffset(),
+            'video_file_chunk' => $chunk->getPartialFile(),
+        ];
+
+        try {
+            $response = $this->sendUploadRequest($endpoint, $params);
+        } catch (FacebookResponseException $e) {
+            $preException = $e->getPrevious();
+            if ($allowToThrow || !$preException instanceof FacebookResumableUploadException) {
+                throw $e;
+            }
+
+            // Return the same chunk entity so it can be retried.
+            return $chunk;
+        }
+
+        return new FacebookTransferChunk($chunk->getFile(), $chunk->getUploadSessionId(), $chunk->getVideoId(), $response['start_offset'], $response['end_offset']);
+    }
+
+    /**
+     * Attempts to upload a chunk of a file in $retryCountdown tries.
+     *
+     * @param string $endpoint
+     * @param FacebookTransferChunk $chunk
+     * @param int $retryCountdown
+     *
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookResponseException
+     */
+    public function maxTriesTransfer($endpoint, FacebookTransferChunk $chunk, $retryCountdown)
```

Comment:
I actually did have this originally in the master `Facebook` class but I was passing around lots of service classes/entities and it seemed like a lot of work of that class. To me it was a smell so I moved it back to the `FacebookResumableUploader`. I don't really see it as syntactic sugar since a resumable uploader seems like a pretty good place to handle retries of a resumable download.

But I'm not 100% married to either way. I'd be happy to move it without too much of a fight. :)